test/units/module_utils/test_facts.py

```diff
@@ -0,0 +1,463 @@
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division)
+__metaclass__ = type
+
+# for testing
+from ansible.compat.tests import unittest
+from ansible.compat.tests.mock import Mock, patch
+
+from ansible.module_utils import facts
+
+HARDWARE_SUBCLASS_MAP = {'Generic': facts.Hardware,
+                         'Linux': facts.LinuxHardware,
+                         'SunOS': facts.SunOSHardware,
+                         'OpenBSD': facts.OpenBSDHardware,
+                         'FreeBSD': facts.FreeBSDHardware,
+                         'DragonFly': facts.DragonFlyHardware,
+                         'NetBSD': facts.NetBSDHardware,
+                         'AIX': facts.AIX,
+                         'HP-UX': facts.HPUX,
+                         'Darwin': facts.Darwin,
+                         }
+
+
+class TestFactsHardwareNew:
+    """Verify that the automagic in Hardware.__new__ selects the right subclass."""
+    @patch('platform.system')
+    def _hardware_new(self, platform_id, hardware_class, mock_platform):
+        mock_platform.return_value = platform_id
+        hardware = facts.Hardware(module=Mock(), load_on_init=False)
+        assert_msg = '%s should have been created for platform=%s' % (hardware_class, platform_id)
+        assert isinstance(hardware, hardware_class), assert_msg
```

Comment:
We get good messages for free if we use unittest.assertIsinstance(), assertEq(), etc.  Need to inherit from unittest.TestCase to have access to those.